crates/fuel-core/src/service/adapters/fuel_gas_price_provider.rs

```diff
@@ -0,0 +1,188 @@
+        Self {
+            totaled_block_height: RefCell::new(block_height),
+            total_reward: RefCell::new(reward),
+            total_cost: RefCell::new(cost),
+        }
+    }
+
+    fn update(&self, block_height: BlockHeight, reward: u64, cost: u64) {
+        let mut totaled_block_height = self.totaled_block_height.borrow_mut();
+        let mut total_reward = self.total_reward.borrow_mut();
+        let mut total_cost = self.total_cost.borrow_mut();
+        while *totaled_block_height < block_height {
+            *totaled_block_height = block_height;
+            *total_reward += reward;
+            *total_cost += cost;
+        }
+    }
+}
+
+impl<FB, DA, A> FuelGasPriceProvider<FB, DA, A>
+where
+    FB: FuelBlockHistory,
+    DA: DARecordingCostHistory,
+    A: GasPriceAlgorithm,
+{
+    fn inner_gas_price(&self, requested_block_height: BlockHeight) -> Result<u64> {
+        let latest_block = self
+            .block_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+        if latest_block > requested_block_height {
+            self.block_history
+                .gas_price(requested_block_height)
+                .map_err(Error::UnableToGetGasPrice)?
+                .ok_or(Error::GasPriceNotFoundForBlockHeight(
+                    requested_block_height,
+                ))
+        } else if Self::asking_for_next_block(latest_block, requested_block_height) {
+            self.calculate_new_gas_price(latest_block)
+        } else {
+            Err(Error::RequestedBlockHeightTooHigh {
+                requested: requested_block_height,
+                latest: latest_block,
+            })
+        }
+    }
+
+    fn asking_for_next_block(
+        latest_block: BlockHeight,
+        block_height: BlockHeight,
+    ) -> bool {
+        *latest_block + 1 == *block_height
+    }
+
+    fn calculate_new_gas_price(&self, latest_block_height: BlockHeight) -> Result<u64> {
+        self.update_totals(latest_block_height)?;
+        let previous_gas_price = self
+            .block_history
+            .gas_price(latest_block_height)
+            .map_err(Error::UnableToGetGasPrice)?
+            .ok_or(Error::GasPriceNotFoundForBlockHeight(latest_block_height))?;
+        let block_fullness = self
+            .block_history
+            .block_fullness(latest_block_height)
+            .map_err(Error::UnableToGetBlockFullness)?
+            .ok_or(Error::BlockFullnessNotFoundForBlockHeight(
+                latest_block_height,
+            ))?;
+        let new_gas_price_candidate = self.algorithm.calculate_gas_price(
+            previous_gas_price,
+            self.total_reward(),
+            self.total_cost(),
+            block_fullness,
+        );
+        let new_gas_price = core::cmp::min(
+            new_gas_price_candidate,
+            self.algorithm.maximum_next_gas_price(previous_gas_price),
+        );
```

Comment:
This is kinda a smell. I'm not sure if this should be in the provider or the algo. I'm getting a feeling like the provider and the algo abstractions could be merged some, especially when we start looking at the estimation trait impl.